File: P3.java
import java.util.*;

// Time Complexity :    O(m x n)
// Space Complexity :   O(m x n)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english

// Your code here along with comments explaining your approach


class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        List<Integer> result = new ArrayList<>();
        int top = 0; int bottom = m -1;
        int left = 0; int right = n - 1;

        while (top <= bottom && left <= right) {
            //top 
            for (int j = left; j <= right; j ++) {
                result.add(matrix[top][j]);
            }
            top ++;

            //right
            for (int i = top; i <= bottom; i ++) {
                result.add(matrix[i][right]);
            }
            right --;

            //bottom
            if (top <= bottom) {
            for (int j = right; j >= left; j --) {
                result.add(matrix[bottom][j]);
            }
            bottom --;
            }

            if (left <= right){
            //left
            for (int i = bottom; i >= top; i --) {
                result.add(matrix[i][left]);
            }
            left ++;
            }
        }
        return result;
    }
}


//recursive soln

// Time Complexity :    O(m x n)
// Space Complexity :   O(m x n)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english

// Your code here along with comments explaining your approach


class Solution2 {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        List<Integer> result = new ArrayList<>();
        int top = 0; int bottom = m-1;
        int left = 0; int right = n-1;
        helper(matrix, result, top, right, left, bottom);
        return result;
    }
    private void helper(int[][] matrix, List<Integer> result, int top, int right, int left, int bottom){
            if(top > bottom || left > right) return;
            // top
            for(int j = left; j <= right; j++){
                result.add(matrix[top][j]);
            }
            top++;
            // right
            for(int i = top; i <= bottom; i++){
                result.add(matrix[i][right]);
            }
            right--;
            
           if(top <= bottom){
            // bottom
                for(int j = right; j >= left; j--){
                    result.add(matrix[bottom][j]);
                }
                bottom--;
           }
            if( left <= right){
                // left
                for(int i = bottom; i >= top; i--){
                    result.add(matrix[i][left]);
                }
                left++;
            }
            helper(matrix, result, top, right, left, bottom);
            
       
    }
}

File: p3.py
#Time COmplexity o(mn)
# Space Complexity O(1)
#Worked on Leet code:yes

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:

        if not matrix or not matrix[0]:
            return []
        left =0
        right = len(matrix[0])-1
        top= 0
        bottom = len(matrix)-1
        res=[]

        while top<=bottom and left<=right:
            for i in range(left, right+1):
                res.append(matrix[top][i])
            top+=1

            for j in range(top,bottom+1):
                res.append(matrix[j][right])
            right-=1

            if top<=bottom:
                for k in range(right,left-1,-1):
                    res.append(matrix[bottom][k])
                bottom-=1

            if left<=right:
                for l in range(bottom, top-1,-1):
                    res.append(matrix[l][left]) 
                left+=1
        return res
